### ISSUE TITLE
Variable length path associations

### ISSUE THREAD

**Issue by cheerfulstoic** [neo4jrb/neo4j#777]
* Support for returning either the last element in the path or the entire path
* Support for defining min/max path length

**Comment by cheerfulstoic** [comment]
Proposed syntax:

```ruby
has_many :out, :name, :type, variable: true

has_many :out, :name, :type, variable: {min: 1, max: 5}
```

**Comment by cheerfulstoic** [comment]
Could call as:

```ruby
object.name.to_a # To get objects at the end
object.name.to_paths # To get the paths

object.name_path # To get path for `has_one`
```

The `QueryProxy#to_paths` could even work in general, perhaps, though we'd need to build one long path as opposed to building each step as a separate part of the `MATCH`

It would also be cool to find a way to turn this into a hierarchy of objects in an eager-loading fashion with just the one query

**Comment by samis** [comment]
Would this work so that you could start from one object and find the path to another object using an association like this one?

**Comment by cheerfulstoic** [comment]
Yeah, it should.  Associations return `QueryProxy` objects (more or less, there's a branch in the works) and `QueryProxy` responds to the `#match_to` method, so you'd be able to say:

```ruby
  source_object.variable_association_name.match_to(target_object).to_paths
```
